File: 57_Insert_Interval.py
class Solution(object):   
    def insert(self, intervals: list, newInterval: list) -> list:
        result = []
        i = 0
        n = len(intervals)
        
        while i < n and intervals[i][1] < newInterval[0]:
            result.append(intervals[i])
            i += 1
            
        # Merge overlapping intervals
        while i < n and intervals[i][0] <= newInterval[1]:
            newInterval[0] = min(newInterval[0], intervals[i][0])
            newInterval[1] = max(newInterval[1], intervals[i][1])
            i += 1
            
        result.append(newInterval)
        
        # Add all remaining intervals after the new interval
        while i < n:
            result.append(intervals[i])
            i += 1
        
        return result
    
if __name__ == "__main__":
    soln = Solution()
    
    intervals = [[1,5]]
    newInterval = [6,8]
    output = soln.insert(intervals, newInterval)
    print(f"list : {output}")
    
    intervals = [[1,5]]
    newInterval = [2,7]
    output = soln.insert(intervals, newInterval)
    print(f"list : {output}")
    
    intervals = [[1,5]]
    newInterval = [2,3]
    output = soln.insert(intervals, newInterval)
    print(f"list : {output}")
    
    intervals = [[1,3],[6,9]]
    newInterval = [2,5]
    output = soln.insert(intervals, newInterval)
    print(f"list : {output}")
    
    intervals = [[1,2],[3,5],[6,7],[8,10],[12,16]]
    newInterval = [4,8]
    output = soln.insert(intervals, newInterval)
    print(f"list : {output}")
    
    intervals = []
    newInterval = [5,7]
    output = soln.insert(intervals, newInterval)
    print(f"list : {output}")

File: 57_Insert_Interval.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> results{};
        int idx = 0;

        while (idx != intervals.size() && intervals[idx][1] < newInterval[0]) {
            results.push_back(intervals[idx]);
            idx += 1;
        }

        while (idx != intervals.size() && intervals[idx][0] <= newInterval[1]) {
            newInterval[0] = min(newInterval[0], intervals[idx][0]);
            newInterval[1] = max(newInterval[1], intervals[idx][1]);
            idx += 1;
        }

        results.push_back(newInterval);

        while (idx != intervals.size()) {
            results.push_back(intervals[idx]);
            idx += 1;
        }

        return results;
    }
};

void print_results(vector<vector<int>>& vec) {
    for (auto _ : vec) {
        for (auto __ : _) {
            cout << __ << " ";
        }
        putchar('\n');
    }
    putchar('\n');
}

int main() {
    Solution soln = Solution();

    vector<vector<int>> intervals = {{1,3}, {6,9}};
    vector<int> newInterval = {2,5};
    auto res = soln.insert(intervals, newInterval);
    print_results(res);

    intervals = {{1,5}};
    newInterval = {2,3};
    res = soln.insert(intervals, newInterval);
    print_results(res);

    intervals = {{1,2}, {3,5}, {6,7}, {8,10}, {12,16}};
    newInterval = {4,8};
    res = soln.insert(intervals, newInterval);
    print_results(res);

    intervals = {};
    newInterval = {5,7};
    res = soln.insert(intervals, newInterval);
    print_results(res);

    return 0;
}